cpp/src/io/parquet/experimental/page_index_filter.cu

```diff
@@ -782,49 +805,40 @@ std::vector<std::vector<bool>> aggregate_reader_metadata::compute_data_page_mask
-        select_page_indices.begin(),
-        is_row_required_fn{row_mask.nullable(), row_mask.null_mask(), row_mask.data<bool>()});
-
-      // Remove duplicate page indices across (presorted) rows
-      auto const filtered_uniq_page_end_iter = thrust::unique(
-        rmm::exec_policy_nosync(stream), select_page_indices.begin(), filtered_pages_end_iter);
-
-      // Number of final filtered pages for this column
-      size_t const num_surviving_pages_this_column =
-        thrust::distance(select_page_indices.begin(), filtered_uniq_page_end_iter);
-
-      total_surviving_pages += num_surviving_pages_this_column;
-
-      // Copy the filtered page indices for this column to host
-      auto host_select_page_indices = cudf::detail::make_host_vector(
-        cudf::device_span<cudf::size_type const>{select_page_indices.data(),
-                                                 num_surviving_pages_this_column},
-        stream);
-
-      // Vector to data page mask the this column
-      auto valid_pages = std::vector<bool>(total_pages_in_this_column, false);
-      std::for_each(host_select_page_indices.begin(),
-                    host_select_page_indices.end(),
-                    [&](auto const page_idx) { valid_pages[page_idx] = true; });
-
-      data_page_mask.push_back(std::move(valid_pages));
+      data_page_mask_tasks.emplace_back(
+        cudf::detail::host_worker_pool().submit_task([&, col_idx = col_idx] {
+          // Construct a row indices mapping based on page row counts and offsets
+          auto const total_pages_this_column = page_row_counts[col_idx].size();
+          auto valid_pages_this_column       = std::vector<bool>(total_pages_this_column, false);
+          // Number of final filtered pages for this column
+          size_t num_surviving_pages_this_column = 0;
+          // For all rows
+          for (auto row_idx = 0; row_idx < total_rows; ++row_idx) {
+            // If this row is required or invalid, add its page index to the output list.
+            if (not is_row_valid[row_idx] or is_row_required[row_idx]) {
+              // binary search to find the page index this row_idx belongs to and set the
+              // page index to true page_indices
+              auto const& offsets = page_row_offsets[col_idx];
+              auto const page_itr = std::upper_bound(offsets.cbegin(), offsets.cend(), row_idx);
+              CUDF_EXPECTS(page_itr != offsets.cbegin() and page_itr != offsets.cend(),
+                           "Invalid page index");
+              auto const page_idx               = std::distance(offsets.cbegin(), page_itr) - 1;
+              valid_pages_this_column[page_idx] = true;
+              num_surviving_pages_this_column++;
+              // Move row_idx to the last row of this page
+              row_idx = offsets[page_idx + 1] - 1;
+            }
+          }
+
+          total_surviving_pages.fetch_add(num_surviving_pages_this_column);
```

Comment:
This work can easily be done in a custom GPU kernel in the future. That way we can also avoid copying `row_mask` to host.